### DISCUSSION TITLE
Support ALL THE LSP

### DISCUSSION THREAD

**Discussion by ChristianChiarulli** [LunarVim/LunarVim#1784] [Ideas]
https://github.com/neovim/nvim-lspconfig/blob/master/CONFIG.md

Everything here should be supported

**Comment by jameswalmsley** [comment]
I was thinking it might be cool if the lsp ftplugins can detect that the server isn't installed,
and then display an "info message" in the status bar. (Something not too intrusive)...

e.g.

Enable language support for C/C++ with: LspInstall cpp

This would prompt new users to get things working.

**Comment by ChristianChiarulli** [comment]
That's a good idea. I'm not sure how to check tho..

**Comment by PZ31k0nauT** [comment]
Maybe this is to simple. Would it be a good start to check if the pre definded installation folder of the given lsp is empty?

**Comment by raprocks** [comment]
we can check for the presence of the LSP by checking if it is executable in path or not. mayvbe that can help

**Comment by kinthaiofficial** [comment]
Supporting all LSP servers in a unified config is a good goal — the key is a declarative approach that doesn't require manual setup for each server.

The `nvim-lspconfig` + `mason.nvim` combination handles most of this:

```lua
-- Automatic LSP setup via mason-lspconfig
return {
  {
    "williamboman/mason.nvim",
    config = true,
  },
  {
    "williamboman/mason-lspconfig.nvim",
    opts = {
      -- Auto-install missing LSPs
      automatic_installation = true,
      ensure_installed = {
        "lua_ls", "pyright", "ts_ls", "rust_analyzer",
        "gopls", "clangd", "jsonls", "yamlls",
      },
    },
  },
  {
    "neovim/nvim-lspconfig",
    config = function()
      local lspconfig = require("lspconfig")
      local capabilities = require("cmp_nvim_lsp").default_capabilities()
      
      -- Default config applied to ALL servers
      local default_opts = {
        capabilities = capabilities,
        on_attach = function(client, bufnr)
          -- Keymaps applied to every LSP buffer
          vim.keymap.set("n", "gd", vim.lsp.buf.definition, { buffer = bufnr })
          vim.keymap.set("n", "K", vim.lsp.buf.hover, { buffer = bufnr })
          vim.keymap.set("n", "<leader>ca", vim.lsp.buf.code_action, { buffer = bufnr })
        end,
      }
      
      -- Server-specific overrides
      local server_opts = {
        lua_ls = {
          settings = { Lua = { diagnostics = { globals = { "vim" } } } }
        },
        rust_analyzer = {
          settings = { ["rust-analyzer"] = { checkOnSave = { command = "clippy" } } }
        },
      }
      
      -- Apply to all installed servers
      require("mason-lspconfig").setup_handlers({
        function(server_name)
          local opts = vim.tbl_deep_extend("force", default_opts, server_opts[server_name] or {})
          lspconfig[server_name].setup(opts)
        end,
      })
    end,
  },
}
```

This pattern means: add a server to `ensure_installed`, and it gets the full default config (keymaps, capabilities, diagnostics) automatically. Server-specific tweaks go in `server_opts`. New LSP support = one line.
